File: app/src/main/java/com/nipa/healthcareapp/NotificationsAdapter.java
package com.nipa.healthcareapp;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.nipa.healthcareapp.data.models.NotificationDB;

import com.google.android.material.card.MaterialCardView;
// com.google.firebase.Timestamp is imported by NotificationDB, but Date and SimpleDateFormat are needed here.
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

public class NotificationsAdapter extends ListAdapter<NotificationDB, NotificationsAdapter.NotificationViewHolder> {

    public NotificationsAdapter() {
        super(new NotificationDiffCallback());
    }

    @NonNull
    @Override
    public NotificationViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_notification, parent, false);
        return new NotificationViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NotificationViewHolder holder, int position) {
        NotificationDB notification = getItem(position);
        if (notification != null) {
            holder.bind(notification);
        }
    }

    static class NotificationViewHolder extends RecyclerView.ViewHolder {
        private final MaterialCardView cardView;
        private final TextView tvTitle;
        private final TextView tvMessage;
        private final TextView tvTime;

        public NotificationViewHolder(@NonNull View itemView) {
            super(itemView);
            cardView = itemView.findViewById(R.id.notificationCard); // Corrected ID from Kotlin source
            tvTitle = itemView.findViewById(R.id.tvNotificationTitle);
            tvMessage = itemView.findViewById(R.id.tvNotificationMessage);
            tvTime = itemView.findViewById(R.id.tvNotificationTime);
        }

        public void bind(NotificationDB notification) {
            tvTitle.setText(notification.getTitle());
            tvMessage.setText(notification.getMessage());

            if (notification.getTimestamp() != null) {
                // Formatting Firebase Timestamp to a more readable format
                Date date = notification.getTimestamp().toDate();
                SimpleDateFormat sdf = new SimpleDateFormat("MMM dd, hh:mm a", Locale.getDefault());
                tvTime.setText(sdf.format(date));
            } else {
                tvTime.setText("");
            }

            // itemView.setOnClickListener(v -> { /* Handle item click if needed */ });
        }
    }

    private static class NotificationDiffCallback extends DiffUtil.ItemCallback<NotificationDB> {
        @Override
        public boolean areItemsTheSame(@NonNull NotificationDB oldItem, @NonNull NotificationDB newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull NotificationDB oldItem, @NonNull NotificationDB newItem) {
            return oldItem.getId().equals(newItem.getId()) &&
                   Objects.equals(oldItem.getTitle(), newItem.getTitle()) &&
                   Objects.equals(oldItem.getMessage(), newItem.getMessage()) &&
                   Objects.equals(oldItem.getTimestamp(), newItem.getTimestamp()) &&
                   Objects.equals(oldItem.getUserId(), newItem.getUserId()) &&
                   oldItem.isRead() == newItem.isRead() &&
                   Objects.equals(oldItem.getType(), newItem.getType()) &&
                   oldItem.isSynced() == newItem.isSynced();
        }
    }
}

File: app/src/main/java/com/nipa/healthcareapp/NotificationsAdapter.kt
package com.nipa.healthcareapp

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.nipa.healthcareapp.data.models.Notification
import com.google.android.material.card.MaterialCardView

class NotificationsAdapter : ListAdapter<Notification, NotificationsAdapter.NotificationViewHolder>(NotificationDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NotificationViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_notification, parent, false)
        return NotificationViewHolder(view)
    }

    override fun onBindViewHolder(holder: NotificationViewHolder, position: Int) {
        val notification = getItem(position)
        holder.bind(notification)
    }

    class NotificationViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val cardView: MaterialCardView = itemView.findViewById(R.id.notificationCard)
        private val titleTextView: TextView = itemView.findViewById(R.id.tvNotificationTitle)
        private val messageTextView: TextView = itemView.findViewById(R.id.tvNotificationMessage)
        private val timeTextView: TextView = itemView.findViewById(R.id.tvNotificationTime)

        fun bind(notification: Notification) {
            titleTextView.text = notification.title
            messageTextView.text = notification.message
            timeTextView.text = notification.timestamp.toDate().toString()
        }
    }

    private class NotificationDiffCallback : DiffUtil.ItemCallback<Notification>() {
        override fun areItemsTheSame(oldItem: Notification, newItem: Notification): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Notification, newItem: Notification): Boolean {
            return oldItem == newItem
        }
    }
}
